# Instagram Contest Winner Picker

The following is a quick web scraping-project, designed with the intention of learning webscraping tools such as Selenium and Beautiful Soup. It also helps to practice some basic regular expresions, lambda expressions and loops.

This is a very simple project but it will require basic python knowledge, patience and browsing skills.


In [1]:
# Link for chrome web driver
# https://chromedriver.storage.googleapis.com/83.0.4103.39/chromedriver_mac64.zip

In [2]:
from selenium import webdriver
from bs4 import BeautifulSoup as bs
import time
import re
from urllib.request import urlopen
import json
from pandas.io.json import json_normalize
import pandas as pd, numpy as np
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC


In [3]:
# check for selenium webdriver version

help(webdriver)

Help on package selenium.webdriver in selenium:

NAME
    selenium.webdriver

DESCRIPTION
    # Licensed to the Software Freedom Conservancy (SFC) under one
    # or more contributor license agreements.  See the NOTICE file
    # distributed with this work for additional information
    # regarding copyright ownership.  The SFC licenses this file
    # to you under the Apache License, Version 2.0 (the
    # "License"); you may not use this file except in compliance
    # with the License.  You may obtain a copy of the License at
    #
    #   http://www.apache.org/licenses/LICENSE-2.0
    #
    # Unless required by applicable law or agreed to in writing,
    # software distributed under the License is distributed on an
    # "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    # KIND, either express or implied.  See the License for the
    # specific language governing permissions and limitations
    # under the License.

PACKAGE CONTENTS
    android (package)
    blackberry (pac

In [ ]:
# Make sure to authorize your machine to open chrome driver

In [54]:
postadd='p/CB_SGeagsVm/' # post URL change it to your desired post
# Gary V post 
browser = webdriver.Chrome(executable_path='./chromedriver')
browser.get('https://www.instagram.com/'+postadd+'')
Pagelength = browser.execute_script("window.scrollTo(0, document.body.scrollHeight);")


In [ ]:
##
##while bool(browser.find_element_by_class_name('dCJp8')) == True:
##   browser.find_element_by_class_name('dCJp8').click()
## time.sleep(3)
   
    

The code in the cell below will use selenium driver to click the "Plus" button on the intagram page to load all the possible comments.

It will take some time to fetch large ammount of comments.

In the post i'm using for testing there are more than 1000 comments and the objective is to fecth all or nearly all of them. 

To run the code in another post, just copy the url from the post and paste it up there.

In [55]:
while bool(browser.find_element_by_class_name('dCJp8')) == True:
    
    wait = WebDriverWait(browser, 30, poll_frequency=1)

    browser.find_element_by_class_name('dCJp8').click()
    
    element = wait.until(EC.element_to_be_clickable((By.CLASS_NAME, "dCJp8")))

TimeoutException: Message: 


The Error above is normal. It is a time out error and it happens because the webdriver could not find the the button according to the conditions we stablished. However even if this error

In [56]:

source = browser.page_source
data=bs(source, 'html.parser')
body = data.find('body')
script = body.find('script', text=lambda t: t.startswith('window._sharedData'))
page_json = script.text.split(' = ', 1)[1].rstrip(';')
data = json.loads(page_json)


In [57]:
target_class = []

participants = []

soup = bs(source,'html.parser')

for link in soup.find_all('a'):
    target_class.append(link.get('class'))
    
for link in soup.find_all('a'):
    if link.get('class') == target_class[0]:
        if link.get('href') != '/ester_exposito/':
            participants.append(link.get('href'))
    

In [58]:
len(participants) # does not get the entire comment section.

926

In [59]:
import random # I will use random to shuffle the usernames since this script
              # was intended as a contest user picker

In [60]:
winner_list = list(participants)
winner_list

['/garyvee/',
 '/garyvee/',
 '/sololifeclothing/',
 '/whereyougoinaj/',
 '/loveyouranxiety/',
 '/tysflight101/',
 '/rezatayebi1/',
 '/recasique/',
 '/littlereddressbook/',
 '/thetopwearcompany/',
 '/therequisitionpage/',
 '/mykhailo_chernei/',
 '/thetopwearcompany/',
 '/kerstin.kermit/',
 '/desi_dj_pro/',
 '/notso_goodart/',
 '/spaitgirl/',
 '/jtb808/',
 '/itsnathaniellackey/',
 '/ny_joemarkrealestate/',
 '/neeshsvegantreats/',
 '/tomski_81/',
 '/ahmedalbadawi__/',
 '/jdmlouis/',
 '/rafrudnicki/',
 '/jackplumley/',
 '/aliammadk/',
 '/yourfavoritedayana/',
 '/millionaireinspired/',
 '/davinderlall/',
 '/mike_gebhardt_/',
 '/the.everyday.millennial/',
 '/015seedz/',
 '/wolfxzeroent/',
 '/binallox/',
 '/ollybrownuk/',
 '/simple_social_success/',
 '/tsgiftsandmore/',
 '/avigpointi/',
 '/sportchat2020/',
 '/iamhenryblessed/',
 '/pal_jaiveer/',
 '/mk.karanouh/',
 '/sarbmotives/',
 '/thememorylanecollective/',
 '/serovius/',
 '/alinateodo/',
 '/feifeiirawan/',
 '/bipolarsuperstar/',
 '/viikas

In [61]:
random.shuffle(winner_list)

In [62]:
winner = random.sample(winner_list,1)

In [63]:
winner

['/bnrowdy/']

In [64]:
win = str(winner)
win = re.sub(r'/',"",win)

In [65]:
print(win)

['bnrowdy']


Monday, June 29 2020 it fetched 926 usernames from a post containing 1000+ comments. The next step will be to tune the wait time of the selenium webdriver, and look to make the code faster and more reliable.


